src/lib/io/sol-gpio-impl-contiki-qmsi.c

```diff
@@ -0,0 +1,350 @@
+    qm_gpio_t port_num;
+    uint8_t pin_num;
+    qm_rc_t ret;
+
+    SOL_LOG_INTERNAL_INIT_ONCE;
+
+#ifndef SOL_NO_API_VERSION
+    if (unlikely(config->api_version != SOL_GPIO_CONFIG_API_VERSION)) {
+        SOL_WRN("Couldn't open gpio that has unsupported version '%u', "
+            "expected version is '%u'",
+            config->api_version, SOL_GPIO_CONFIG_API_VERSION);
+        return NULL;
+    }
+#endif
+
+    port_num = GPIO_GET_PORT(pin);
+    SOL_INT_CHECK(port_num, >= NUM_GPIO_PORTS, NULL);
+
+    port = &ports[port_num];
+
+    pin_num = GPIO_GET_PIN(pin);
+    SOL_INT_CHECK(pin_num, >= port->config->num_pins, NULL);
+
+    if (port->opened_pins & BIT(pin_num)) {
+        SOL_WRN("GPIO pin %d of port %u is already open", pin_num, port_num);
+        return NULL;
+    }
+
+    ret = qm_gpio_get_config(port_num, &port_cfg);
+    SOL_EXP_CHECK(ret != QM_RC_OK, NULL);
+
+    gpio = calloc(1, sizeof(*gpio));
+    SOL_NULL_CHECK(gpio, NULL);
+
+    gpio->pin = pin;
+    gpio->active_low = config->active_low;
+
+    mask = BIT(pin_num);
+
+    if (config->dir == SOL_GPIO_DIR_IN) {
+        gpio->cb = config->in.cb;
+        gpio->cb_data = config->in.user_data;
+
+        port_cfg.direction &= ~mask;
+
+        if (config->in.trigger_mode == SOL_GPIO_EDGE_NONE)
+            port_cfg.int_en &= ~mask;
+        else {
+            int iret;
+
+            port_cfg.int_en |= mask;
+            iret = port->config->setup_isr(&port_cfg);
+            SOL_INT_CHECK_GOTO(iret, < 0, setup_isr_error);
+            iret = sol_ptr_vector_append(&port->registered_irqs, gpio);
+            SOL_INT_CHECK_GOTO(iret, < 0, append_error);
+
+            if (config->in.trigger_mode == SOL_GPIO_EDGE_BOTH)
+                port_cfg.int_bothedge |= mask;
+            else {
+                uint8_t polarity = 0;
+
+                port_cfg.int_bothedge &= ~mask;
+                port_cfg.int_type |= mask;
+
+                if (config->in.trigger_mode == SOL_GPIO_EDGE_RISING)
+                    polarity = 1;
+                polarity ^= gpio->active_low;
+
+                if (polarity)
+                    port_cfg.int_polarity |= mask;
+                else
+                    port_cfg.int_polarity &= ~mask;
+            }
+        }
+    } else {
+        port_cfg.direction |= mask;
+        port_cfg.int_en &= ~mask;
+    }
+
+    ret = qm_gpio_set_config(port_num, &port_cfg);
+    SOL_EXP_CHECK_GOTO(ret != QM_RC_OK, set_config_error);
+
+    port->opened_pins |= mask;
+
+    return gpio;
+
+set_config_error:
+    sol_ptr_vector_remove(&port->registered_irqs, gpio);
+setup_isr_error:
+append_error:
+    free(gpio);
+    return NULL;
+}
+
+void
+sol_gpio_close(struct sol_gpio *gpio)
+{
+    struct gpio_port *port;
+    qm_gpio_port_config_t port_cfg;
+    uint32_t mask;
+    qm_gpio_t port_num;
+    uint8_t pin_num;
+    qm_rc_t ret;
+
+    SOL_NULL_CHECK(gpio);
+
+    port_num = GPIO_GET_PORT(gpio->pin);
+    pin_num = GPIO_GET_PORT(gpio->pin);
```

Comment:
On Thu, 19 May 2016 07:55:30 -0700, Vinicius Gomes wrote:
> > +}
> > +
> > +void
> > +sol_gpio_close(struct sol_gpio *gpio)
> > +{
> > +    struct gpio_port *port;
> > +    qm_gpio_port_config_t port_cfg;
> > +    uint32_t mask;
> > +    qm_gpio_t port_num;
> > +    uint8_t pin_num;
> > +    qm_rc_t ret;
> > +
> > +    SOL_NULL_CHECK(gpio);
> > +
> > +    port_num = GPIO_GET_PORT(gpio->pin);
> > +    pin_num = GPIO_GET_PORT(gpio->pin);
> 
> `GPIO_GET_PIN()`

Oops... clearly a function we never use in those systems.